include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,261 @@
+#else
+#define TRITONCACHE_DECLSPEC
+#define TRITONCACHE_ISPEC
+#endif
+#else
+#if defined(_MSC_VER)
+#define TRITONCACHE_DECLSPEC __declspec(dllimport)
+#define TRITONCACHE_ISPEC __declspec(dllexport)
+#else
+#define TRITONCACHE_DECLSPEC
+#define TRITONCACHE_ISPEC
+#endif
+#endif
+
+struct TRITONCACHE_Cache;
+struct TRITONCACHE_CacheEntry;
+struct TRITONCACHE_CacheEntryItem;
+
+///
+/// TRITONCACHE API Version
+///
+/// The TRITONCACHE API is versioned with major and minor version
+/// numbers. Any change to the API that does not impact backwards
+/// compatibility (for example, adding a non-required function)
+/// increases the minor version number. Any change that breaks
+/// backwards compatibility (for example, deleting or changing the
+/// behavior of a function) increases the major version number. A
+/// cache implementation should check that the API version used to compile
+/// the cache is compatible with the API version of the Triton server
+/// that it is running in. This is typically done by code similar to
+/// the following which makes sure that the major versions are equal
+/// and that the minor version of Triton is >= the minor version used
+/// to build the cache.
+///
+///   uint32_t api_version_major, api_version_minor;
+///   TRITONCACHE_ApiVersion(&api_version_major, &api_version_minor);
+///   if ((api_version_major != TRITONCACHE_API_VERSION_MAJOR) ||
+///       (api_version_minor < TRITONCACHE_API_VERSION_MINOR)) {
+///     return TRITONSERVER_ErrorNew(
+///       TRITONSERVER_ERROR_UNSUPPORTED,
+///       "triton cache API version does not support this cache");
+///   }
+///
+#define TRITONCACHE_API_VERSION_MAJOR 0
+#define TRITONCACHE_API_VERSION_MINOR 1
+
+/// Get the TRITONCACHE API version supported by Triton. This
+/// value can be compared against the
+/// TRITONCACHE_API_VERSION_MAJOR and
+/// TRITONCACHE_API_VERSION_MINOR used to build the cache to
+/// ensure that Triton is compatible with the cache.
+///
+/// \param major Returns the TRITONCACHE API major version supported
+/// by Triton.
+/// \param minor Returns the TRITONCACHE API minor version supported
+/// by Triton.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_ApiVersion(
+    uint32_t* major, uint32_t* minor);
+
+///
+/// Cache Lifetime Management
+///
+
+/// Create a new cache object. The caller takes ownership of the
+/// TRITONCACHE_Cache object and must call
+/// TRITONCACHE_CacheDelete to release the object.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache Returns the new cache object.
+/// \param config The config options to initialize the cache with.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheNew(
+    TRITONCACHE_Cache** cache, TRITONSERVER_Message* config);
```

Comment:
I don't think the type of `config` matters, Triton itself doesn't need to know anything about the config, so you can even pass a buffer (void* buf, uint32_t byte_size) that is read directly from the config file. The cache implementation will have complete control over how the config is formatted (can be JSON, protobuf, or custom key-value list)